### ISSUE TITLE
`final val`s that is implemented with a literal constant false positive in initialization checker

### ISSUE THREAD

**Issue by nicolasstucki** [lampepfl/dotty#12769]
## Compiler version

3.0.0

## Semantics of `final val` with literal constants

All `final val`s that is implemented with a literal constant do not generate a field in the bytecode. Their accessors simply return the value and the field is elided.

This optimization makes the bytecode smaller and helps the JVM JIT to inline this value.

## Initialization checker

The initialization checker is currently unaware of how code is generated for these fields. It believes that we do generate the fields and that there is a potential for uninitialized values. The reality is that these will always be seen as initialized as they always return the correct value.

We need to teach the initialization checker about this field encoding.

**Comment by nicolasstucki** [comment]
@liufengyun could you point to the place where this should be done?

**Comment by liufengyun** [comment]
> @liufengyun could you point to the place where this should be done?

It can be easily done below with an additional check to skip the warning if the `sym.info` is a constant type :

https://github.com/lampepfl/dotty/blob/45e2ad7d12f4fbe2637e4e5405cf15ad84df0a1f/compiler/src/dotty/tools/dotc/transform/init/Semantic.scala#L290-L292

I am just not sure that whether it's a good thing. From personal experience, if I were a programmer, I don't want my program semantics to change by adding an explicit type annotation `Int`. I would be happy to receive a warning like this.
In most cases, the user can re-order the fields to fix the warning, and they will not think the warning is annoying. 

However, there are cases where the warning cannot be easily fixed (the compiler contains 2 instances related to `def maxToken`), which is annoying for advanced users who explicitly employ the language feature.

/cc: @olhotak @EnzeXing WDYT?

**Comment by liufengyun** [comment]
The `maxToken` example I'm talking about is the following in the compiler:

```Scala
abstract class TokensCommon {
  def maxToken: Int

  val tokenString, debugString: Array[String] = new Array[String](maxToken + 1)

  def enter(token: Int, str: String, debugStr: String = ""): Unit = {
    assert(tokenString(token) == null)
    tokenString(token) = str
    debugString(token) = if (debugStr.isEmpty) str else debugStr
  }
}

object Tokens extends TokensCommon { 
   final val minToken = EMPTY  
  final def maxToken: Int = XMLSTART

  final val COLONEOL = 88;         enter(COLONEOL, ":", ": at eol")
  final val SELFARROW = 89;        enter(SELFARROW, "=>") // reclassified ARROW following self-type

  /** XML mode */
  final val XMLSTART = 99;         enter(XMLSTART, "$XMLSTART$<") // TODO: deprecate
}
```

> However, there are cases where the warning cannot be easily fixed (the compiler contains 2 instances related to `def maxToken`), which is annoying for advanced users who explicitly exploit the language feature.


For those advanced users, maybe they should use `@unchecked`, as they know what they are doing.

**Comment by olhotak** [comment]
It appears that the semantics that the compiler should implement are still being debated in https://github.com/lampepfl/dotty/issues/12570.

Once the semantics are decided there, my opinion is that the initialization checker should follow the actual semantics as closely as possible rather than trying to also be a linter, trying to enforce a "better", "cleaner" semantics than the compiler actually implements. If something is safe but only due to ugly hacks, I think the initialization checker should not report it unsafe. Many use the type checker as an oracle to improve their understanding of the specification of the type system. Similarly, many will use the initialization checker to improve their understanding of the details of initialization. Inconsistency between the init checker and the actual init semantics will be confusing to that understanding.

It sounds like the #12570 is converging towards preserving the semantics of `final` for backward compatibility with Scala 2 and Java but preferring `inline` to better express the intended semantics. In that case, there could be a lint-style deprecation/migration warning, separate from the initialization checker, that suggests the use of `inline` instead of `final`.

In summary, as a general rule, I think the initialization checker should match actual semantics as closely as possible and lint/style warnings should be separate and marked as such without suggesting that a runtime error will occur.

**Comment by dwijnand** [comment]
Does everyone agree with @sjrd, (from https://github.com/lampepfl/dotty/issues/12570#issuecomment-857805134) that

> The init checker should not be changed; it is correct to flag any use-before-initialization as unspecified behavior.

?

**Comment by liufengyun** [comment]
I think we can close the issue, given that no one disagree with @sjrd https://github.com/lampepfl/dotty/issues/12570#issuecomment-857805134.
